### PR TITLE
Wire the inclusion list store into BeaconChain

### PR DESCRIPTION
**Pull request by rahulbarmann**
## Issue Addressed

Part of FOCIL (EIP-7805). This puts the inclusion list store from #9744 onto `BeaconChain` and adds the read wrappers that will be used by gossip verification, fork-choice enforcement, and block production.

## Proposed Changes

* Add an `inclusion_list_store` field to `BeaconChain`, initialise it from `ChainSpec` in `builder.rs`, and prune it in `per_slot_task`.
* Add `inclusion_list_committee`, which resolves the ordered committee for a slot and returns it together with the dependent root used to key the store.
* Add `get_inclusion_list_transactions`, `get_inclusion_list_bits`, and `is_inclusion_list_bits_inclusive` as wrappers around the store reads.
* Add `BeaconChainError::InclusionListStoreError`.
* Add tests covering each read path, the epoch boundary case, and pruning through `per_slot_task`.

This is independent of #9846, and nothing is calling these wrappers yet. Gossip verification will be the follow-up.

## Additional Info

The wrappers take an anchor block root rather than using the head. `with_committee_cache` only resolves a block's own or next epoch shuffling, while inclusion list reads use `slot - 1`. Once per epoch, that falls into the previous epoch and fails when using the head root. There's a test covering this case.

`with_committee_cache` already gives us the attester shuffling decision root, so we get both the committee and the store key from the same lookup. This is also the root that the producer puts in the message as `dependent_root`.

`get_inclusion_list_transactions` only needs to resolve the shuffling for the dependent root and then discards the committee, since that root comes out of the same lookup.

### CODE REVIEW

File: beacon_node/beacon_chain/tests/inclusion_list_store.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            .is_inclusion_list_bits_inclusive(block_root, slot, &bits, false)
+            .unwrap()
+    );
+    assert!(
+        !harness
+            .chain
+            .is_inclusion_list_bits_inclusive(block_root, slot, &timely_bits, false)
+            .unwrap()
+    );
+
```
Comment (conache) [inline beacon_node/beacon_chain/tests/inclusion_list_store.rs:178]: nit: we should also add an assertion using `only_timely=true` for completeness. Mistakenly changing the `only_timely` value to false in `chain.is_inclusion_list_bits_inclusive` would still make the test pass

```suggestion
assert!(
        harness
            .chain
            .is_inclusion_list_bits_inclusive(block_root, slot, &timely_bits, true)
            .unwrap()
    );
```

**Review by conache — COMMENTED** [review]
Looking good to me, I just left a nit comment

File: beacon_node/beacon_chain/tests/inclusion_list_store.rs (hunk 1 above)
Comment (rahulbarmann) [reply beacon_node/beacon_chain/tests/inclusion_list_store.rs:184]: nice catch, thanks! applied it in 506c387db714e4e70608db4e141a1a15f30b920e

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 2]
```diff
@@ -7333,6 +7338,86 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         )
     }
 
+    /// The ordered inclusion list committee for `slot`, and the dependent root that keys it.
+    ///
+    /// `block_root` must not be from a later epoch than `slot`, since `with_committee_cache` cannot
+    /// resolve a shuffling earlier than its own block's epoch.
+    pub fn inclusion_list_committee(
+        &self,
+        block_root: Hash256,
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/beacon_chain.rs]: can we rename this arg from `block_root` to `parent_block_root` so that its clear which root we are supposed to pass into this function

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 3]
```diff
@@ -7333,6 +7338,86 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         )
     }
 
+    /// The ordered inclusion list committee for `slot`, and the dependent root that keys it.
+    ///
+    /// `block_root` must not be from a later epoch than `slot`, since `with_committee_cache` cannot
+    /// resolve a shuffling earlier than its own block's epoch.
+    pub fn inclusion_list_committee(
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/beacon_chain.rs:7357]: we should mention in the doc comment that this function takes a fork choice read lock (via `with_committee_cache`)

File: beacon_node/beacon_chain/src/beacon_chain.rs [hunk 4]
```diff
@@ -7333,6 +7338,86 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
...
+                    .committee_cache
+                    .get_inclusion_list_committee_at_slot(slot, committee_size)?
+                    .into_iter()
+                    .map(|index| index as u64)
+                    .collect::<Vec<_>>();
+
+                Ok((FixedVector::new(committee)?, dependent_root))
+            },
+        )
+    }
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/beacon_chain.rs:7379]: we might not need the committee cache. `get_inclusion_list_transactions` only needs the dependent root.

I think we could do something like

```rust
let shuffling_epoch = slot.epoch(T::EthSpec::slots_per_epoch());
let parent_block = self
    .canonical_head
    .fork_choice_read_lock()
    .get_block(&parent_block_root)
    .ok_or(Error::MissingBeaconBlock(parent_block_root))?;

BlockShufflingIds {
    current: parent_block.current_epoch_shuffling_id.clone(),
    next: parent_block.next_epoch_shuffling_id.clone(),
    previous: None,
    block_root: parent_block.root,
}
.id_for_epoch(shuffling_epoch)
.map(|id| id.shuffling_decision_block)
.ok_or(Error::InvalidShufflingId {
    shuffling_epoch,
    head_block_epoch: parent_block.slot.epoch(T::EthSpec::slots_per_epoch()),
})
```

This avoids the case where a cache miss happens in `with_commitee_cache` (which takes the shuffling write lock, loads a state from disk, advances it and builds the committee cache).

**Review by eserilev — CHANGES_REQUESTED** [review]
looks good on the whole, just a few things

Also:

`get_inclusion_list_commitee` returns a raw `FixedVector<u64, E::InclusionListCommitteeSize>`. We can replace that with the type alias you introduced `InclusionListCommittee`

We should move this type alias `InclusionListCommittee` to the types crate as well

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 2 above)
Comment (rahulbarmann) [reply beacon_node/beacon_chain/src/beacon_chain.rs]: addressed in a37adf66604873b34e02dd8448e490cb5e4c8890

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 3 above)
Comment (rahulbarmann) [reply beacon_node/beacon_chain/src/beacon_chain.rs:7357]: addressed in a37adf66604873b34e02dd8448e490cb5e4c8890

File: beacon_node/beacon_chain/src/beacon_chain.rs (hunk 4 above)
Comment (rahulbarmann) [reply beacon_node/beacon_chain/src/beacon_chain.rs:7379]: thanks! good call. also turns out my tests weren't actually covering the dependent root resolution. swapping the shuffling decision root for the parent root still made everything pass, since `bits_and_transactions_read_back_through_the_chain` runs at genesis where both roots are the same. I've added assertions to the previous-epoch test which catch it now.

**Comment by rahulbarmann** [comment]
Thanks for the review @eserilev! I've addressed all of them in a37adf66604873b34e02dd8448e490cb5e4c8890

Should be good to take another look whenever you get the chance, thanks!

**Review by eserilev — APPROVED** [review]
LGTM
